### ISSUE TITLE
Consumer inside of Navigator page rebuilds after being removed

### ISSUE THREAD

**Issue by ElteHupkes** [rrousselGit/river_pod#1169]
**Describe the bug**
Let me preface this by saying that I don't know if this is something that RiverPod can do anything about or something I'll just have to work around, but I recently ran into the following situation:

- I have a `Navigator` with a list of pages that depends on a `StateNotifier`
- One of these pages is included as long as a certain piece of information is present in this notifier
- That page performs some kind of logic to display this information when it's present. It `ref.watch()`es the information for this purpose. This logic isn't safe to perform if the information isn't present. This shouldn't be a problem, because if the information isn't present, the page isn't shown.
- *However*, when the information is removed, despite the page no longer being in the list, it is rebuilt once - resulting in an error being thrown.

If that explanation is too confusing, here's a concrete example:
- A profile page is shown in a `Navigator` as long as there's a user profile in the app state.
- The profile page prints the user name, thereby assuming it isn't `null` (e.g. `ref.watch(appStateProvider).profile!.name`), because if it were `null`, the page wouldn't be shown.
- The profile page also contains a button to log out, which sets the profile to `null` in the `StateNotifier`.
- When pressing this button, the profile page is rebuilt _after_ the rebuild of the `Navigator`, which no longer includes the page. The profile is now `null` and throws on `profile!.name`.

I suspect this has something to do with `Navigator` keeping the old widget around for comparison when building up its new list, because if I try to recreate the issue with a `Column` it doesn't happen. I'm not sure if this is something that could be detected.

**To Reproduce**
Here's a somewhat convoluted example:
```dart
import 'package:flutter/material.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends ConsumerWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final state = ref.watch(aProvider);

    return MaterialApp(
      title: 'Flutter Demo',
      home: Navigator(
        pages: [
          const MaterialPage(child: WidgetA(), name: 'a'),
          if (state.showValue) const MaterialPage(child: WidgetB(), name: 'b'),
        ],
        onPopPage: (route, _) => true,
      ),
    );
  }
}

class WidgetA extends ConsumerWidget {
  const WidgetA({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          onPressed: () =>
              ref.read(aProvider.notifier).state = const MyState(10),
          child: const Text('Show value screen'),
        ),
      ),
    );
  }
}

class WidgetB extends ConsumerWidget {
  const WidgetB({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('Value is ${ref.watch(aProvider).value!}'),
            ElevatedButton(
              onPressed: () =>
                  ref.read(aProvider.notifier).state = const MyState(null),
              child: const Text('Hide value screen'),
            )
          ],
        ),
      ),
    );
  }
}

class MyState {
  final int? value;
  bool get showValue => value != null;

  const MyState(this.value);
}

final a = StateController<MyState>(const MyState(10));
final aProvider =
    StateNotifierProvider<StateController<MyState>, MyState>((ref) => a);
```

Note that:
- If `MyState.value == null`, the page containing `WidgetB` is no longer a part of the widget tree.
- Despite this, when pressing `Hide value screen` setting `value=null`, `WidgetB` will still build once, and throw.

**Expected behavior**
`WidgetB` is no longer specified as part of the widget tree and thus shouldn't rebuild. 

Like I said at the beginning, I don't know if RiverPod has control over this or it's something to be sorted out in `Navigator`. If there's any further investigation I can do, please let me know.

**Comment by lucavenir** [comment]
Hi, I had a similar problem in the past.

I also posted an issue, which answers in here about this behavior that should be useful to you: #943 

TLDR: Unluckily, this behavior is intended. Check @rrousselGit 's [reply](https://github.com/rrousselGit/river_pod/issues/943#issuecomment-980578434).

I'd ask you to not close this issue since I actually want to hear from Remi or someone else what's the **best practice** when it comes to Routing and Riverpod altogether. I think it's for the best to have documentation on this matter. This is especially important since Routing is strictly connected to State (thus Router 2.0).

For the record I still have heavy problems with Navigation, Providers re-building themselves and Asynchronous operations linked within the two.

**Comment by ElteHupkes** [comment]
Ah, thanks @lucavenir, that's basically the exact same issue indeed. I did look for similar issues but it looks my my search terms were off. I can see how this is hard for RiverPod to do anything about, but it doesn't seem _fundamentally_ unsolvable; maybe we should point our arrows at Flutter / Navigator.

Navigation (in particular Navigator 2.0) is a bit of an issue indeed, as evidenced by [several](https://github.com/rrousselGit/river_pod/issues/1023) [issues](https://github.com/rrousselGit/river_pod/issues/946). I'm guessing that one of the issues is that `RouterDelegate` is explicitly a `ChangeNotifier`, a pattern which [Remi doesn't want to promote](https://github.com/rrousselGit/river_pod/issues/1023#issuecomment-1024895682), but since it's the only way to recalculate some of the router state I don't see any way around at least using _some_ `ChangeNotifier` functionality. A bit of a weird choice on account of the Flutter team perhaps.

I do have this wired up by basically just adding a listener to a `StateNotifier` do keep that behavior contained within the router, but it's kind of ugly, and this particular issue is obviously still an issue.

So perhaps any example is a bad example at the moment, and the best practice may not be all that good. I would agree with you that an example of how to do it "not terribly" would probably be valuable.

**Comment by ElteHupkes** [comment]
For what it's worth, this is my current workaround:

```dart
class Cache<T> {
  T? value;
  Cache();

  T get(T? latest) => (value = latest ?? value)!;
}

final profileCacheProvider = Provider.autoDispose((ref) => Cache<Profile>());
final safeProfileProvider = Provider.autoDispose(
  (ref) => ref.watch(profileCacheProvider).get(ref.watch(maybeProfileProvider)),
  dependencies: [
    maybeProfileProvider,
    profileCacheProvider,
  ],
);
```

The `maybeProfileProvider` returns a `Profile?` which is always non-null when this provider is first used. It updates the latest value in a container as long as it's non null. When the problematic call happens, it uses the last known value one last time before being disposed.

**Comment by rrousselGit** [comment]
Since the problem was explained, what is expected from this issue?

**Comment by lucavenir** [comment]
I'd need to know how to properly work with Riverpod and Router 2.0 (or even 1.0) without unexpected rebuilds of the Providers linked between to stack pages. I'm missing the best practice and all I can find on the internet / docs / issues in here are workarounds.

As mentioned in #943 I think I'll open a new issue addressing this with reproducible examples that mess up my production app without posting my whole app of course. But again this issue is not mine so I'm just thinking out loud at this point :-p

**Comment by ElteHupkes** [comment]
I think @lucavenir wanted me to keep it open to draw attention to the confusion surrounding combining the Router API with RiverPod. That being said, I actually meant to close it after my last comment, because I believe other issues exist for this already. So I'll close it now.

It's a shame that the only solution I'm currently aware of is a somewhat hacky workaround. I do not believe simply accepting that a widget may rebuild after it has logically been removed is a satisfactory conclusion. It seems to me that it's deemed to be an issue with Navigator though, not RiverPod, which is fair enough.

Thanks for your time, Remi. 98% of the time RiverPod is a great help, you just happen to be on the receiving end of the 2% when I need to bitch about it a little. Just wanted to point that out 😉.

**Comment by ElteHupkes** [comment]
Oh, just to add some random extra information - I ran into [this issue on StackOverflow](https://stackoverflow.com/questions/63848523/flutter-navigator-pop-running-widget-build-method-again?rq=1) earlier today, which from reading it seems to be a very similar issue (ignoring the somewhat absurd premise and accepted answer for a second). `ModalRoute.of(context)` probably triggers a rebuild on the removed widget when the route stack changes. No RiverPod involved, just `InheritedWidget`. So yeah, we're barking up the wrong tree...
